In [3]:
import pandas as pd
import numpy as np
import re

In [4]:
data = pd.read_csv('data/ProjectOpenRefine.csv')

In [5]:
print(data.columns)

Index(['FMID', 'MarketName', 'Website', 'Facebook', 'Twitter', 'TwitterHandle',
       'Youtube', 'OtherMedia', 'Instagram', 'street', 'city', 'County',
       'State', 'zip', 'Season1Date', 'Season1Time', 'Season2Date',
       'Season2Time', 'Season3Date', 'Season3Time', 'Season4Date',
       'Season4Time', 'x', 'y', 'Location', 'Credit', 'WIC', 'WICcash',
       'SFMNP', 'SNAP', 'Organic', 'Bakedgoods', 'Cheese', 'Crafts', 'Flowers',
       'Eggs', 'Seafood', 'Herbs', 'Vegetables', 'Honey', 'Jams', 'Maple',
       'Meat', 'Nursery', 'Nuts', 'Plants', 'Poultry', 'Prepared', 'Soap',
       'Trees', 'Wine', 'Coffee', 'Beans', 'Fruits', 'Grains', 'Juices',
       'Mushrooms', 'PetFood', 'Tofu', 'WildHarvested', 'updateTime'],
      dtype='object')


In [6]:
pd.set_option('display.max_columns', None)
data.head()

FMID                                       MarketName  \
0  1018261  Caledonia Farmers Market Association - Danville   
1  1018318                Stearns Homestead Farmers' Market   
2  1009364                106 S. Main Street Farmers Market   
3  1010691              10th Steet Community Farmers Market   
4  1002454                             112st Madison Avenue   

                                             Website  \
0  https://sites.google.com/site/caledoniafarmers...   
1                    http://www.StearnsHomestead.com   
2             http://thetownofsixmile.wordpress.com/   
3                                                NaN   
4                                                NaN   

                                            Facebook Twitter TwitterHandle  \
0  https://www.facebook.com/Danville.VT.Farmers.M...     NaN           NaN   
1                      StearnsHomesteadFarmersMarket     NaN           NaN   
2                                                NaN     NaN           NaN   
3                                                NaN     NaN           NaN   
4                                                NaN     NaN           NaN   

  Youtube                                         OtherMedia Instagram  \
0     NaN                                                NaN       NaN   
1     NaN                                                NaN       NaN   
2     NaN                                                NaN       NaN   
3     NaN  http://agrimissouri.com/mo-grown/grodetail.php...       NaN   
4     NaN                                                NaN       NaN   

                   street      city     County           State    zip  \
0                     NaN  Danville  Caledonia         Vermont   5828   
1         6975 Ridge Road     Parma   Cuyahoga            Ohio    NaN   
2      106 S. Main Street  Six Mile        NaN  South Carolina  29682   
3  10th Street and Poplar     Lamar     Barton        Missouri  64759   
4    112th Madison Avenue  New York   New York        New York  10029   

                Season1Date                                   Season1Time  \
0  06/14/2017 to 08/30/2017                         Wed: 9:00 AM-1:00 PM;   
1  06/24/2017 to 09/30/2017                         Sat: 9:00 AM-1:00 PM;   
2                       NaN                                           NaN   
3  04/02/2014 to 11/30/2014    Wed: 3:00 PM-6:00 PM;Sat: 8:00 AM-1:00 PM;   
4          July to November  Tue:8:00 am - 5:00 pm;Sat:8:00 am - 8:00 pm;   

                Season2Date            Season2Time Season3Date Season3Time  \
0  09/06/2017 to 10/18/2017  Wed: 2:00 PM-6:00 PM;         NaN         NaN   
1                       NaN                    NaN         NaN         NaN   
2                       NaN                    NaN         NaN         NaN   
3                       NaN                    NaN         NaN         NaN   
4                       NaN                    NaN         NaN         NaN   

  Season4Date Season4Time          x          y                      Location  \
0         NaN         NaN -72.140337  44.411036                           NaN   
1         NaN         NaN -81.733939  41.374801                           NaN   
2         NaN         NaN -82.818700  34.804200                           NaN   
3         NaN         NaN -94.274619  37.495628                           NaN   
4         NaN         NaN -73.949300  40.793900  Private business parking lot   

  Credit WIC WICcash SFMNP SNAP Organic Bakedgoods Cheese Crafts Flowers Eggs  \
0      Y   Y       N     Y    N       Y          Y      Y      Y       Y    Y   
1      Y   N       N     Y    N       -          Y      N      Y       Y    Y   
2      Y   N       N     N    N       -        NaN    NaN    NaN     NaN  NaN   
3      Y   N       N     N    N       -          Y      N      Y       N    Y   
4      N   N       Y     Y    N       -          Y      N      Y       Y    N   

  Seafood Herbs Vegetables Hone

### Twitter Handle Clean Up
Below I wrote a function that was mapped onto the TwitterHandle column that was previously created in OpenRefine
This did some further refining to isolate only the twitter handle in this column

In [7]:
def twitter_function(handle):
    """
    Function used to map the twitter handles values
    to raw @ token, this done by stripping the 
    http address execept for the @handle
    """
    http_find = re.split("htts://twitter.com/",handle)
    https_find = re.split("https://twitter.com/",handle)
    www_find =  re.split("www.twitter.com/",handle)
    if len(http_find) > 1:
        return(http_find
               [-1])
    elif len(https_find) > 1:
        return (https_find[-1])
    elif len(www_find) > 1:
        return(www_find[-1])
    else:
        return handle

#Run the function
data['TwitterHandle'] = data['TwitterHandle'].map(twitter_function,na_action = 'ignore')

Below is the number of missing entries in each column 

In [8]:
len(data) -data.count()

FMID                0
MarketName          0
Website          3475
Facebook         4758
Twitter          7679
TwitterHandle    7683
Youtube          8516
OtherMedia       7968
Instagram        8190
street            290
city               40
County            515
State               0
zip               945
Season1Date      3208
Season1Time      3050
Season2Date      8238
Season2Time      8250
Season3Date      8606
Season3Time      8610
Season4Date      8681
Season4Time      8681
x                  29
y                  29
Location         5958
Credit              0
WIC                 0
WICcash             0
SFMNP               0
SNAP                0
                 ... 
Bakedgoods       2933
Cheese           2933
Crafts           2933
Flowers          2933
Eggs             2933
Seafood          2933
Herbs            2933
Vegetables       2933
Honey            2933
Jams             2933
Maple            2933
Meat             2933
Nursery          2933
Nuts             2933
Plants    

## GeoLocation Clean up
The main use case we are interested relies heavily on the accuracy of the geolocation data of this dataset. Although the dates maybe good enough to locate most of the farmers markets any gains that can be made in accuracy will greatly impact the efficacy of the clustering used in the specific use case discussed above. With this is mind I decided to focus on the NaN values that pertain the geolocation. 

As can be seen above there is an inconsistent amount of information missing in the address location for each farmers market:

* x            =      29
* y            =     29
* street       =     290
* city         =      40
* County       =     515
* State        =       0
* zip          =     945

The list above is increasing in granularity of geolocation. We can use external information on zipcodes, countys to populate some of the missing information



The information supplementation will come from a [Census Dataset](https://simplemaps.com/data/us-cities) which containes the zipcode, city, county information for cities around the country


Also the Census dataset uses state codes so I needed [this](https://gist.github.com/rogerallen/1583593) python dictonary to get the state codes in a hash map

In [9]:
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'District of Columbia': 'DC',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Northern Mariana Islands':'MP',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Palau': 'PW',
    'Pennsylvania': 'PA',
    'Puerto Rico': 'PR',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virgin Islands': 'VI',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY',
}

In [10]:
#create new column with codes
data['StateCode'] = data['State'].apply(lambda x: us_state_abbrev[x])

In [11]:
#All data elements where the city is unknown and the Postalcode is known which appears there are none which makes sense
data.loc[data['city'].isnull() & data['zip'].notnull()]

#All data elements where the county is unknown and the city is known
countyNoZip = data.loc[data['County'].isnull() & data['zip'].notnull()].index.values.tolist()

In [14]:
#Read in CSV file based on Census data to provide supplemental geolocation information
us_cities = pd.read_csv('data/uscitiesv1.5.csv')

In [15]:
#If the City/State pair is found in the census dataset and the county is unknown we replace the entries for the county information in the orginal dataframe
for index in countyNoZip:
    try:
        data.loc[index,'County'] = us_cities.loc[(us_cities['city'] == data.loc[index,'city']) & (us_cities['state_id'] == data.loc[index,'StateCode']),'county_name'].values[0]
    except IndexError:
        pass

### GeoLocation Results

We were able to go from 515 unknown countys to 141 which means that 374 counties were identfied using the supplemntal database

### Further data curation

* One can do some more refining on the dataset to try and identifiy more counties but for our usage this is no helpful since we have the X,Y coordinate data.
* There are  2933 missing food information rows in this data set which compromises 33% of the dataset. 
I orginally had thought that this was related to the update time and 
that all entries updated after a certain date would not have missing information but after 
doing the following exploration below I can see that thhere is indeed no coorelation 
between the update year and wheter or not is including food values
for my particular use case the geoloaction data and the food data are most important. 
Therfore for the schema I create and the database I populate
I will use pandas dropna function to remove all rows that do not have food related 
information and all entries that do not at least have X,Y information.

In [16]:
years = data.loc[data['Wine'].isnull(),'updateTime'].map(lambda x : x[0:4] )

In [17]:
year_values = {}

In [18]:
for row in years:
    if row not in year_values:
        year_values[row] = '0'

In [19]:
year_values

{'2013': '0',
 '2011': '0',
 '2009': '0',
 '2014': '0',
 '2010': '0',
 '2012': '0',
 '2015': '0',
 '2016': '0',
 'May ': '0',
 '2017': '0',
 '5/14': '0',
 '7/29': '0',
 'Apr ': '0',
 '4/4/': '0',
 '4/29': '0',
 'Jul ': '0',
 'Jun ': '0',
 '2/7/': '0',
 '3/17': '0',
 '8/29': '0',
 '6/4/': '0',
 '10/1': '0',
 '6/12': '0',
 '4/21': '0',
 '4/22': '0',
 '4/14': '0',
 '1/14': '0',
 '8/26': '0',
 '6/28': '0',
 '8/7/': '0'}

### Drop Columns
For our final use case these are the columns where we are dropping rows that have NaN values in these columns:
* x
* y
* X food values





In [20]:
 data = (data[data['y'].notnull() | data['x'].notnull()])

In [21]:
data = (data[data['Bakedgoods'].notnull()])

In [22]:
data = data.drop('Twitter',axis=1)

In [24]:
data.to_csv('data/clean_data.csv')